syntax = "proto3";

service File {
  rpc read(ReadFileReq) returns (ReadFileRsp) {}
  rpc write(WriteFileReq) returns (WriteRsp) {}
}
message ReadFileRsp {
  bytes buffer = 1;
}

message ReadFileReq {
  string filename = 1;
  string chunkname = 2;
}

message WriteFileReq {
  string filename = 1;
  string chunkname = 2;
  bytes buffer = 3;
}
message WriteRsp {
}

service NameNodeService {
    rpc open(FileOpenReq) returns (stream DatanodeList) {}
    rpc create(FileCreateReq) returns (stream DatanodeList){}
    rpc heart_beat(DatanodeInfo) returns (Empty){}
    rpc report(ChunkReport) returns (Empty) {}
    rpc listin(Empty) returns (stream DirectoryContent) {}
}

message FileOpenReq {
  string filename = 1;
}
message FileCreateReq{
    string filename = 1;
    int64 chunks_number = 2;
}

message DatanodeList{
    string localization = 1;
    string chunkname= 2;

}
message Empty{}

message DatanodeInfo{
  string id = 1;
  string socket = 2;
  bool is_leader = 3;
}

message ChunkReport{
  string partname = 1;
  string filename = 2;
  string location = 3;
}

message DirectoryContent {
  string name = 1;
}